### PR TITLE
add improvements on subscription with activation key

### PR DESCRIPTION
**Pull request by adeliaferreira**
# Description

After hot fixes from last week, this PR will address some of comments we have open on subscription to image. 
It's create a new func that will create the subscription object to fill "req.customizations.subscription" if adequate and more tests to validate it.

FIXES: 

## Type of change

What is it?

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation update
- [ ] Tests update
- [ ] Refactor

### CODE REVIEW

File: pkg/clients/imagebuilder/client.go [hunk 1]
```diff
@@ -347,22 +367,13 @@ func (c *Client) ComposeInstaller(image *models.Image) (*models.Image, error) {
...
-			Insights:      true,
-			RHC:           true,
-		}
-
+	subscription, error := createSubscriptionToImage(image)
+	if error != nil {
+		c.log.WithField("error", error.Error()).Error("Error can not convert org Id to integer")
+		return nil, error
 	}
+	req.Customizations.Subscription = subscription
```
Comment (omaciel) [inline pkg/clients/imagebuilder/client.go:375]: If image.ActivationKey is empty, then the new function `createSubscriptionToImage` will return a `nil` value for `Subscription`. Is it okey to assign `nil` to `req.Customizations.Subscription` or should we just not added it?

**Review by omaciel — COMMENTED** [review]
I love that you created a helper function! Left a question for you. I was a bit confused with the tests, but I don't have time to review the entire test suite (maybe some of the context I'm missing is already set somewhere else?).

File: pkg/clients/imagebuilder/client_test.go [hunk 2]
```diff
@@ -830,6 +830,83 @@ var _ = Describe("Image Builder Client Test", func() {
 		Expect(img.Commit.ExternalURL).To(BeFalse())
 	})
 
+	It("test compose installer with activation key", func() {
+		// enable feature flag
+		err := os.Setenv(feature.PassUserToImageBuilder.EnvVar, "true")
```
Comment (ldjebran) [inline pkg/clients/imagebuilder/client_test.go:835]: we do not need this feature flag

File: pkg/clients/imagebuilder/client_test.go [hunk 3]
```diff
@@ -830,6 +830,83 @@ var _ = Describe("Image Builder Client Test", func() {
...
+		installer := models.Installer{Username: faker.Username(), SSHKey: faker.UUIDHyphenated()}
+		composeJobID := "compose-job-id-returned-from-image-builder"
+		ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			b, err := io.ReadAll(r.Body)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(b).ToNot(BeNil())
+			var req ComposeRequest
+			err = json.Unmarshal(b, &req)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(req.Customizations.Subscription.ActivationKey).To(Equal("test-key"))
```
Comment (ldjebran) [inline pkg/clients/imagebuilder/client_test.go:847]: we check the other subscription variables

File: pkg/clients/imagebuilder/client_test.go [hunk 4]
```diff
@@ -830,6 +830,83 @@ var _ = Describe("Image Builder Client Test", func() {
...
+		img, err = client.ComposeInstaller(img)
+		Expect(err).ToNot(HaveOccurred())
+		Expect(img).ToNot(BeNil())
+		Expect(img.Installer.ComposeJobID).To(Equal(composeJobID))
+		Expect(img.Commit.ExternalURL).To(BeFalse())
+	})
+
+	It("test compose installer with no activation key", func() {
+		// enable feature flag
+		err := os.Setenv(feature.PassUserToImageBuilder.EnvVar, "true")
```
Comment (ldjebran) [inline pkg/clients/imagebuilder/client_test.go:874]: we do not need this feature flag

File: pkg/clients/imagebuilder/client.go (hunk 1 above)
Comment (ldjebran) [reply pkg/clients/imagebuilder/client.go:375]: req.Customizations.Subscription is nil by initialization , it's a pointer 
```go
// Customizations is made of the packages that are baked into an image
type Customizations struct {
	Packages            *[]string     `json:"packages"`
	PayloadRepositories *[]Repository `json:"payload_repositories,omitempty"`
	Users               []User        `json:"users,omitempty"`
	Subscription        *Subscription `json:"subscription,omitempty"`
}
```
so it's OK to assign a nil value

File: pkg/clients/imagebuilder/client_test.go (hunk 2 above)
Comment (adeliaferreira) [reply pkg/clients/imagebuilder/client_test.go:835]: removed, thanks for the comment

File: pkg/clients/imagebuilder/client_test.go (hunk 3 above)
Comment (adeliaferreira) [reply pkg/clients/imagebuilder/client_test.go:844]: Added on last commit

File: pkg/clients/imagebuilder/client_test.go (hunk 4 above)
Comment (adeliaferreira) [reply pkg/clients/imagebuilder/client_test.go]: removed

File: pkg/clients/imagebuilder/client_test.go (hunk 3 above)
Comment (ldjebran) [reply pkg/clients/imagebuilder/client_test.go:844]: eg before accessing Subscription variables you have to check that it's not nil

```go
Expect(req.Customizations.Subscription).ToNot(BeNil())
```

File: pkg/clients/imagebuilder/client_test.go [hunk 5]
```diff
@@ -830,6 +830,82 @@ var _ = Describe("Image Builder Client Test", func() {
...
+		composeJobID := "compose-job-id-returned-from-image-builder"
+		ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			b, err := io.ReadAll(r.Body)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(b).ToNot(BeNil())
+			var req ComposeRequest
+			err = json.Unmarshal(b, &req)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(req.Customizations.Subscription.ActivationKey).To(Equal("test-key"))
+			Expect(req.Customizations.Subscription.Organization).To(Equal(234))
```
Comment (ldjebran) [inline pkg/clients/imagebuilder/client_test.go:845]:
```suggestion
			Expect(req.Customizations.Subscription.Organization).To(Equal(orgID))
```

File: pkg/clients/imagebuilder/client_test.go [hunk 6]
```diff
@@ -830,6 +830,82 @@ var _ = Describe("Image Builder Client Test", func() {
...
+			Expect(req.Customizations.Subscription.RHC).To(Equal(true))
+			Expect(req.Customizations.Subscription.Insights).To(Equal(true))
+			Expect(req.Customizations.Subscription.BaseUrl).To(Equal(config.Get().SubscriptionBaseUrl))
+			Expect(req.Customizations.Subscription.ServerUrl).To(Equal(config.Get().SubscriptionBaseUrl))
```
Comment (ldjebran) [inline pkg/clients/imagebuilder/client_test.go:849]:
```suggestion
			Expect(req.Customizations.Subscription.ServerUrl).To(Equal(config.Get().SubscriptionServerURL))
```
Think also we should initialize these config values at the top of this test, and restore them in AfterEach function

File: pkg/clients/imagebuilder/client_test.go [hunk 7]
```diff
@@ -73,6 +76,10 @@ var _ = Describe("Image Builder Client Test", func() {
 		client = InitClient(context.Background(), log.NewEntry(log.StandardLogger()))
 		// save the original image builder url
 		originalImageBuilderURL = conf.ImageBuilderConfig.URL
+		err = os.Setenv(SUBSCRIPTION_BASE_URL, "http://base.url")
+		Expect(err).ToNot(HaveOccurred())
+		err = os.Setenv(SUBSCRIPTION_SERVER_URL, "http://server.url")
+		Expect(err).ToNot(HaveOccurred())
```
[the remark this answers is not in the document]
Comment (adeliaferreira) [reply pkg/clients/imagebuilder/client_test.go]: done

**Review by ldjebran — APPROVED** [review]
lgtm, many thanks Tay

**Review by de1987 — APPROVED** [review]

**Pull request merged**